src/coreclr/nativeaot/Runtime/unix/PalRedhawkInline.h

```diff
@@ -87,14 +87,26 @@ FORCEINLINE int64_t PalInterlockedCompareExchange64(_Inout_ int64_t volatile *pD
     return result;
 }
 
-#if defined(HOST_AMD64) || defined(HOST_ARM64)
+#if defined(HOST_AMD64) || defined(HOST_ARM64) || defined(HOST_LOONGARCH64)
 FORCEINLINE uint8_t PalInterlockedCompareExchange128(_Inout_ int64_t volatile *pDst, int64_t iValueHigh, int64_t iValueLow, int64_t *pComparandAndResult)
 {
+#if defined(HOST_LOONGARCH64)
+    int64_t iResult0 = __sync_val_compare_and_swap(pDst, pComparandAndResult[0], iValueLow);
+    int64_t iResult1 = __sync_val_compare_and_swap(pDst+1, pComparandAndResult[1], iValueHigh);
```

Comment:
This needs to swap both values atomically. This implementation is not correct.

Is there no 128-bit CAS on Loongarch?

Reply:
Yes, there is no 128-bit CAS on LoongArch, and I agree that it cannot be written like this.
So can LoongArch not add this interface?
Thanks!

Reply:
I thought LoongArch 1.1 has the `sc.q` instruction that serves this purpose. Am I wrong, or is it irrelevant because it's too new instruction?

Reply:
@sunlijun-610, could you try:

```c++
    __int128_t iComparand = ((__int128_t)pComparandAndResult[1] << 64) + (uint64_t)pComparandAndResult[0];
    __int128_t expected = iComparand;
    __int128_t newValue = ((__int128_t)iValueHigh << 64) + (uint64_t)iValueLow;
    uint8_t success = __atomic_compare_exchange_n((__int128_t volatile*)pDst, &expected, newValue, false, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST);
    PalInterlockedOperationBarrier();
    pComparandAndResult[0] = (int64_t)expected;
    pComparandAndResult[1] = (int64_t)(expected >> 64);
    return success;
```

It should work for all archs except for HOST_X86, which requires libatomic (https://github.com/dotnet/runtime/pull/10287).

Reply:
I checked about atomic operations, and found there's no 128-bit CAS on LoongArch. Maybe not supported physically.
Thus cannot use `__atomic_compare_exchange_n` , or it will report errors like `large atomic operation may incur significant performance penalty; the access size (16 bytes) exceeds the max lock-free size (8  bytes)`.
Is there any other way to support it, or can we just not use this interface?
Thanks a lot!

Reply:
I checked about atomic operations, and found there's no 128-bit CAS on LoongArch. Maybe not supported physically.
Thus cannot use `__atomic_compare_exchange_n` , or it will report errors like `large atomic operation may incur significant performance penalty; the access size (16 bytes) exceeds the max lock-free size (8  bytes)`.
Is there any other way to support it, or can we just not use this interface?
Thanks a lot!

Reply:
You would have to refactor the callers of this method to not require 128-bit atomic operations.

Reply:
We can also use mutex/lock solution under `#ifdef TARGET_LOONGARCH64`: https://godbolt.org/z/MhGWh48sG

Reply:
We can also use mutex/lock solution under `#ifdef TARGET_LOONGARCH64`: https://godbolt.org/z/6r1rnjfzd

Reply:
AFAIK there's a single use of `PalInterlockedCompareExchange128` in `UpdatePointerPairAtomically` to update interface dispatch cache. In that particular case it's possible to use the double-word CAS (DWCAS) emulation algorithms that use sentinel values and/or abuse unused bits of the pointer (eg. bottom bits from alignment, top bits unused in the 2^40/2^48 address space).

I'm not going to go into too much details because at least two variants of this DWCAS emulations are patented, and I am not sure if the patents have expired or if there was some explicitly granted rights for them. They have different properties in relation to preventing stalls, ie. one thread being de-scheduled in the middle of emulated DWCAS may or may not prevent a different thread from succeeding depending on particular algorithm.

Reply:
It would be best to omit this change from this PR and work on it separately.

Reply:
OK, I will consider more about this.
I will revert this change in this PR first, and later create a new PR for this issue.
Thanks a lot for helping!